File: parallelprograming/haskell/Matrix.hs
module Matrix where

import Data.Array
import Data.List(groupBy, intercalate)
import Data.Function(on)
import Control.Parallel.Strategies(using, parList, rpar)
import Control.DeepSeq

type IntArray2 = Array (Int, Int)

listIntArray2 upBounds = listArray (botBounds, upBounds)
  where botBounds = (1,1) :: (Int, Int)

newtype Matrix a = Matrix (IntArray2 a)

instance (Num a, NFData a) => Num (Matrix a) where
  (+) = indOp (+)
  (-) = indOp (-)
  (*) = matrixOp (\a b -> 
          let (_, (n1, m1)) = bounds a
              (_, (n2, m2)) = bounds b
              resMat = [ let resSum = sum [ a ! (i, k) * b ! (k, j) 
                                          | k <- [1..m1]
                                          ]
                         in resSum `seq` resSum
                       | i <- [1..n1], j <- [1..m2]
                       ]
          in if m1 == n2
             then listIntArray2 (n1, m2) resMat
             else error "Matrix.(*) first width must be the same as second height"
        )
  fromInteger x = Matrix $ listIntArray2 (1, 1) [fromInteger x]
  abs _ = error "Matrix.abs"
  signum _ = error "Matrix.signum"
  
matrixOp f (Matrix a) (Matrix b) = Matrix $ f a b
matmap f (Matrix a) = Matrix $ f a

indOp f = matrixOp $ (\a b -> 
    let bounds_a = bounds a
        bounds_b = bounds b
    in if bounds_a == bounds_b
       then listArray bounds_a $ zipWith f (elems a) (elems b)
       else error "indOp: sizes must be the same"
  )

instance Show a => Show (Matrix a) where
  show (Matrix arr) = intercalate "\n" . map ( intercalate "\t" 
                                             . map (show . snd)
                                             ) 
                    . groupBy ((==) `on` fst.fst) 
                    $ assocs arr

File: parallelprograming/java/Matrix.java
import java.util.Random;

public class Matrix {
  private int[][] elems;
  private Matrix(int[][] elems) {
    this.elems = elems;
  }

  public Matrix mult(Matrix vec) {
    if (elems.length == 1 && elems[0].length == 1) {
      int k = elems[0][0];
      int[][] nElems = new int[vec.elems.length][vec.elems[0].length];
      for (int i = 0; i < nElems.length; i++) {
        for (int j = 0; j < nElems[0].length; j++) {
          nElems[i][j] *= k;
        }
      }
      return new Matrix(nElems);
    }
    if (vec.elems.length == 1 && vec.elems[0].length == 1) {
      return vec.mult(this);
    }
    if (elems[0].length != vec.elems.length) {
      throw new IllegalArgumentException("First matrix width must be the same as second matrix height");
    }
    int[][] nElems = new int[elems.length][vec.elems[0].length];
    for (int i = 0; i < elems.length; i++) {
      for (int j = 0; j < vec.elems[0].length; j++) {
        nElems[i][j] = 0;
        for (int k = 0; k < elems[0].length; k++) {
          nElems[i][j] += elems[i][k] * vec.elems[k][j];
        }
      }
    }
    return new Matrix(nElems);
  }

  public Matrix trans() {
    int[][] nElems = new int[elems[0].length][elems.length];
    for (int i = 0; i < nElems.length; i++) {
      for (int j = 0; j < nElems[0].length; j++) {
        nElems[i][j] = elems[j][i];
      }
    }
    return new Matrix(nElems);
  }

  public Matrix max() {
    int max = elems[0][0];
    for (int i = 0; i < elems.length; i++) {
      for (int j = 0; j < elems[0].length; j++) {
        max = elems[i][j] > max ? elems[i][j] : max;
      }
    }
    int[][] res = new int[1][1];
    res[0][0] = max;
    return new Matrix(res);
  }

  public void print() {
    for (int i = 0; i < elems.length; i++) {
      for (int j = 0; j < elems[0].length; j++) {
        System.out.print(elems[i][j]);
        System.out.print("\t");
      }
      System.out.println();
    }
  }

  public static Matrix F1(Matrix b, Matrix c, Matrix ma, Matrix me) {
    return b.mult(c).mult(ma.mult(me)); 
  }

  public static Matrix F2(Matrix mk, Matrix mh, Matrix mf) {
    return mk.trans().mult(mh.mult(mf));
  }

  public static Matrix F3(Matrix mp, Matrix mr, Matrix t) {
    return mp.mult(mr).max().mult(t);
  }

  public static Matrix randomMatrix(int m, int n, int maxVal) {
    int[][] elems = new int[m][n];
    for (int i = 0; i < m; i++) {
      elems[i] = randomArray(n, maxVal);
    }
    return new Matrix(elems);
  }

  public static Matrix randomVVector(int size, int maxVal) {
    int[][] res = new int[1][];
    res[0] = randomArray(size, maxVal);
    return new Matrix(res);
  }

  public static Matrix randomHVector(int size, int maxVal) {
    int[][] res = new int[size][1];
    int[] elems = randomArray(size, maxVal);
    for (int i = 0; i < res.length; i++) {
      res[i][0] = elems[i];
    }
    return new Matrix(res);
  }

  private static int[] randomArray(int size, int maxVal) {
    Random rand = new Random();
    int[] elems = new int[size];
    for (int i = 0; i < size; i++) {
      elems[i] = rand.nextInt(maxVal);
    }
    return elems;
  }

}
